### PR TITLE
feat: move-stable row ids in compaction

### PR DESCRIPTION
**Pull request by wjones127**
Part of #2307.

Also addresses #2397

### CODE REVIEW

File: rust/lance/src/dataset/optimize/remapping.rs
Comment (wjones127) [inline rust/lance/src/dataset/optimize/remapping.rs:1]: These functions are moved from the parent module.

File: rust/lance-table/src/rowids.rs [hunk 1]
```diff
@@ -152,6 +152,27 @@ impl RowIdSequence {
         self.0.extend_from_slice(remaining_segments);
     }
 
+    /// Mask
```
Comment (westonpace) [inline rust/lance-table/src/rowids.rs:155]:
```suggestion
    /// Delete row ids by position
```
Or some other comment explaining the difference between this and `delete` (they look very similar)

File: rust/lance/src/dataset/optimize.rs [hunk 2]
```diff
@@ -839,6 +696,54 @@ async fn rewrite_files(
...
+            .position(|frag| frag.id as u32 == *frag_id)
+            .expect("Fragment not found")
+    });
+
+    // Need to remove deleted rows
+    futures::stream::iter(old_sequences.iter_mut().zip(old_fragments.iter()))
+        .map(Ok)
+        .try_for_each(|((_, seq), frag)| async move {
+            let deletions = read_deletion_file(&dataset.base, frag, dataset.object_store()).await?;
+            if let Some(deletions) = deletions {
```
Comment (westonpace) [inline rust/lance/src/dataset/optimize.rs:720]: Are we guaranteed the deletion was materialized at this point?  Is there any way two fragments can be compacted without materializing their deletions?

File: rust/lance/src/dataset/optimize.rs [hunk 3]
```diff
@@ -865,26 +773,42 @@ pub async fn commit_compaction(
...
+        // In absence of remapping, we did not reserve fragment ids as part of
+        // the individual rewrite tasks. This is partly to avoid the commit
+        // conflict bug described in https://github.com/lancedb/lance/issues/2397
+        // and partly because we don't need to reserve fragment ids if we're not
+        // remapping indices.
+        let new_fragments = rewrite_groups
+            .iter_mut()
+            .flat_map(|group| group.new_fragments.iter_mut())
+            .collect::<Vec<_>>();
+        reserve_fragment_ids(dataset, new_fragments.into_iter()).await?;
```
Comment (westonpace) [inline rust/lance/src/dataset/optimize.rs:809]: Do we even need to call `reserve_fragment_ids` in this case?  For some reason I thought commit would just assign ids if they were left as -1.

**Review by westonpace — APPROVED** [review]

File: rust/lance/src/dataset/optimize.rs (hunk 3 above)
Comment (wjones127) [reply rust/lance/src/dataset/optimize.rs:809]: That's a good point. We don't need to create another transaction for that.

File: rust/lance/src/dataset/optimize.rs (hunk 2 above)
Comment (wjones127) [reply rust/lance/src/dataset/optimize.rs:720]: Yeah that is guaranteed. We compacting by scanning and then writing out. The scan removes deleted rows.

File: rust/lance/src/dataset/optimize.rs (hunk 3 above)
Comment (wjones127) [reply rust/lance/src/dataset/optimize.rs]: Nevermind, turns out this was in fact load bearing. The assigned fragment ids from this step are used to compute the new fragment bitmap for each of the indices.

Comment (westonpace) [reply rust/lance/src/dataset/optimize.rs:806]: Ah, thanks for the try

**Pull request merged**